### DISCUSSION TITLE
How to specify a microphone device and redirect it for USB camera?

### DISCUSSION THREAD

**Discussion by shmerl** [FreeRDP/FreeRDP#8562] [Q&A]
I'm trying to connect with wlfreerdp to Windows host and managed USB for camera, but I can't make microphone to work (I already enabled Allow audio recording redirection in group policy editor).

The issue is that when I redirect the camera's USB device, microphone seems to disappear from the client side Linux system (may be becasue of that very USB redirection), becasue microphone is part of the USB connected camera itself.

What is the right way to do it? And is there in general a way to specify the microphone device if there are multiple ones? (As a workaround I can try getting a standalone microphone).

**Comment by shmerl** [comment]
OK, found this bug: https://github.com/FreeRDP/FreeRDP/issues/6024

So after all it's not possible to use the microphone from the camera itself. I'll try to get a separate one. But then the second question still applies. What if the client system has several microphone devices (not part of the camera). Is there a way to specify which one freerdp should use?

**Reply by akallabeth** [reply]
depends on your backend, but it uses the backends default if you specify nothing.
if you need to use a specific one for some reason, you can add `/microphone:dev:<somedevicename>` where `<somedevicename>` is a name the backend understands (so check your `pulse` , `alsa`, `oss` or whatever you have for these

**Reply by shmerl** [reply]
Is that the correct syntax? How it will differentiate between alsa and pulse devices for example?

With `pactil list` I get for instance this for one of my microphones:

```
device.name = "alsa_card.usb-Shure_Inc_Shure_MV7-00"
```

But using `/microphone:dev:alsa_card.usb-Shure_Inc_Shure_MV7-00` didn't seem to help and I also don't see such device mentioned anywhere if debug logging is enabled.

**Reply by shmerl** [reply]
Hm, it might be related to this: https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/756

What would be the syntax for alsa device?

**Reply by shmerl** [reply]
OK, after a lot of trial and error, I managed to make it work using such syntax!

```
... /microphone:sys:alsa,dev:hw:0,0 ...
```
It's one of the syntaxes for ALSA devices. indexes 0,0 can be determined with:

```
arecord -l
**** List of CAPTURE Hardware Devices ****
card 0: MV7 [Shure MV7], device 0: USB Audio [USB Audio]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
...
```

Note card 0, device 0 in my case, so that would make `hw:0,0`. So anyone trying to set up their microphone, pick accordingly.

Question remains still what syntax should be used for pulse devices naming.

Also, I suppose pipewire sys / device is a current missing piece in freerdp.

**Reply by shmerl** [reply]
One problem that I see with that redirection is that microphone has almost 4 seconds delay! Is there any way to improve that?

**Reply by shmerl** [reply]
Found [this thread](https://github.com/FreeRDP/FreeRDP/issues/4791) and suggestion to use `format:1`. It indeed improved latency, but introduced horrible noise to the audio.

Another observation - initially the lag isn't very big, but the more the microphone is used, the bigger the lag becomes.

Also, when this lag happens, such errors show in the terminal:

```
ALSA lib pcm.c:8570:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8570:(snd_pcm_recover) underrun occurred
...
```

**Comment by shmerl** [comment]
I tried various names for pulse device, but nothing really works in that mode. Can you please clarify what's the syntax for it is?

And as alternative, would redirecting a USB audio device directly work? I tried doing it with my USB microphone, but that didn't get detected on the Windows side.

I see errors like:

```
 libusb_control_transfer: error LIBUSB_ERROR_PIPE[-9]
```

Though such error happens when redirecting a webcam too and that works fine, so it's probably not the reason.

**Comment by akallabeth** [comment]
* redirecting audio devices with USB redirection is not supported by windows RDP server
* I´d leave the backend to default (mostly `pulse`) as `alsa` is very prone to noise if the callbacks are not processed fast enough
* also, you really need to specify your microphone device directly? why not just use the default (e.g. the microphone set up as default input of your system)?
* adding the freerdp version you try with might help

**Comment by shmerl** [comment]
When I'm trying simply default pulse, audio from the microphone isn't picked up.

It might be because my setup is a bit complicated.

I have 3 audio inputs on the Linux side:

* Motherboard audio in (nothing is attached to it, but it exists as alsa device).
* USB webcam (that's being redirected for webcam purposes).
* Shure MV7 USB mic.

When redirection happens, webcam disappears from the mics that pulse / pipewire detects. But somehow the result still isn't picking audio form MV7.

It might be hitting this bug (but I don't see any errors in the freerdp log): https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/756

My version is 2.9.0 from Debian testing:

```
wlfreerdp /buildconfig
This is FreeRDP version 2.9.0 (2.9.0)
Build configuration: BUILD_TESTING=OFF BUILTIN_CHANNELS=ON HAVE_AIO_H=1 HAVE_EXECINFO_BACKTRACE=1 HAVE_EXECINFO_BACKTRACE_SYMBOLS=1 HAVE_EXECINFO_BACKTRACE_SYMBOLS_FD=1 HAVE_EXECINFO_H=ON HAVE_EXECINFO_HEADER=1 HAVE_FCNTL_H=1 HAVE_GETLOGIN_R=1 HAVE_GETPWUID_R=1 HAVE_INTTYPES_H=1 HAVE_JOURNALD_H=TRUE HAVE_MATH_C99_LONG_DOUBLE=1 HAVE_POLL_H=1 HAVE_PTHREAD_MUTEX_TIMEDLOCK=ON HAVE_PTHREAD_MUTEX_TIMEDLOCK_LIBS= HAVE_PTHREAD_MUTEX_TIMEDLOCK_SYMBOL=1 HAVE_SYSLOG_H=1 HAVE_SYS_EVENTFD_H=1 HAVE_SYS_FILIO_H= HAVE_SYS_MODEM_H= HAVE_SYS_SELECT_H=1 HAVE_SYS_SOCKIO_H= HAVE_SYS_STRTIO_H= HAVE_SYS_TIMERFD_H=1 HAVE_TM_GMTOFF=1 HAVE_UNISTD_H=1 HAVE_XI_TOUCH_CLASS=1 WITH_ALSA=ON WITH_CAIRO=OFF WITH_CCACHE=ON WITH_CHANNELS=ON WITH_CLANG_FORMAT=ON WITH_CLIENT=ON WITH_CLIENT_AVAILABLE=1 WITH_CLIENT_CHANNELS=ON WITH_CLIENT_CHANNELS_AVAILABLE=1 WITH_CLIENT_COMMON=ON WITH_CLIENT_INTERFACE=OFF WITH_CUPS=ON WITH_DEBUG_ALL=OFF WITH_DEBUG_CAPABILITIES=OFF WITH_DEBUG_CERTIFICATE=OFF WITH_DEBUG_CHANNELS=OFF WITH_DEBUG_CLIPRDR=OFF WITH_DEBUG_DVC=OFF WITH_DEBUG_KBD=OFF WITH_DEBUG_LICENSE=OFF WITH_DEBUG_MUTEX=OFF WITH_DEBUG_NEGO=OFF WITH_DEBUG_NLA=OFF WITH_DEBUG_NTLM=OFF WITH_DEBUG_RAIL=OFF WITH_DEBUG_RDP=OFF WITH_DEBUG_RDPDR=OFF WITH_DEBUG_RDPEI=OFF WITH_DEBUG_RDPGFX=OFF WITH_DEBUG_REDIR=OFF WITH_DEBUG_RFX=OFF WITH_DEBUG_RINGBUFFER=OFF WITH_DEBUG_SCARD=OFF WITH_DEBUG_SND=OFF WITH_DEBUG_SVC=OFF WITH_DEBUG_SYMBOLS=OFF WITH_DEBUG_THREADS=OFF WITH_DEBUG_TIMEZONE=OFF WITH_DEBUG_TRANSPORT=OFF WITH_DEBUG_TSG=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF=OFF WITH_DEBUG_TSMF_AVAILABLE=0 WITH_DEBUG_URBDRC=OFF WITH_DEBUG_WND=OFF WITH_DEBUG_X11=OFF WITH_DEBUG_X11_CLIPRDR=OFF WITH_DEBUG_X11_LOCAL_MOVESIZE=OFF WITH_DEBUG_XV=OFF WITH_DSP_EXPERIMENTAL=OFF WITH_DSP_FFMPEG=ON WITH_EVENTFD_READ_WRITE=1 WITH_FAAC=OFF WITH_FAAD2=OFF WITH_FFMPEG=TRUE WITH_FFMPEG=TRUE WITH_GFX_H264=ON WITH_GPROF=OFF WITH_GSM=ON WITH_GSSAPI=OFF WITH_ICU=ON WITH_INTERNAL_MD4=OFF WITH_INTERNAL_MD5=OFF WITH_IPP=OFF WITH_JPEG=ON WITH_LAME=OFF WITH_LIBRARY_VERSIONING=ON WITH_LIBSYSTEMD=ON WITH_MACAUDIO=OFF WITH_MACAUDIO=OFF WITH_MACAUDIO_AVAILABLE=0 WITH_MANPAGES=ON WITH_MBEDTLS=OFF WITH_OPENCL=OFF WITH_OPENH264=OFF WITH_OPENSLES=OFF WITH_OPENSSL=ON WITH_OSS=ON WITH_PAM=ON WITH_PCSC=ON WITH_PROFILER=OFF WITH_PROXY=OFF WITH_PULSE=ON WITH_SAMPLE=OFF WITH_SANITIZE_ADDRESS=OFF WITH_SANITIZE_ADDRESS_AVAILABLE=1 WITH_SANITIZE_MEMORY=OFF WITH_SANITIZE_MEMORY_AVAILABLE=1 WITH_SANITIZE_THREAD=OFF WITH_SANITIZE_THREAD_AVAILABLE=1 WITH_SERVER=ON WITH_SERVER_CHANNELS=ON WITH_SERVER_INTERFACE=ON WITH_SHADOW=ON WITH_SMARTCARD_INSPECT=OFF WITH_SOXR=OFF WITH_SSE2=ON WITH_SWSCALE=ON WITH_THIRD_PARTY=OFF WITH_VAAPI=OFF WITH_VALGRIND_MEMCHECK=OFF WITH_VALGRIND_MEMCHECK_AVAILABLE=1 WITH_VERBOSE_WINPR_ASSERT=ON WITH_WAYLAND=ON WITH_WINPR_TOOLS=ON WITH_X11=ON WITH_XCURSOR=ON WITH_XDAMAGE=ON WITH_XEXT=ON WITH_XFIXES=ON WITH_XI=ON WITH_XINERAMA=ON WITH_XKBFILE=ON WITH_XRANDR=ON WITH_XRENDER=ON WITH_XSHM=ON WITH_XTEST=ON WITH_XV=ON WITH_ZLIB=ON
Build type:          RelWithDebInfo
CFLAGS:              -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -Wall -Wno-unused-result -Wno-unused-but-set-variable -Wno-deprecated-declarations -fvisibility=hidden -Wimplicit-function-declaration -Wredundant-decls -g -fno-omit-frame-pointer -DWINPR_DLL
Compiler:            GNU, 12.2.0
Target architecture: x64
```

Here is a snippet from `pactl list`:

```
Source #52
	State: SUSPENDED
	Name: alsa_input.usb-046d_C922_Pro_Stream_Webcam_D7C1577F-02.analog-stereo
	Description: C922 Pro Stream Webcam Analog Stereo
	Driver: PipeWire
...
Source #54
	State: SUSPENDED
	Name: alsa_input.usb-Shure_Inc_Shure_MV7-00.mono-fallback
	Description: Shure MV7 Mono
	Driver: PipeWire
...
```

Does debug output show which pulse device is picked for audin?

**Reply by akallabeth** [reply]
this sounds more like your local system has a problem.
I´m constantly switching between laptop internal and my usb headset even in running sessions, so please check that you can actually use the microphone locally properly

**Comment by shmerl** [comment]
I just tried it with explicitly setting my MV7 microphone in KDE as "record all audio via this device" in audio settings. Also, in audacity when recording with default pulse setting it's being picked up OK.

But when used with wlfreerdp with simply `/microphone`, I see in the log that pulseaudio is picked for audin, but it's not showing which particular device, and then in Windows when testing remote audio in the sound setting, it reports that 0% sound was max. I don't see anything anomalous in the debug log output either. There is clearly some kind of disconnect in the audio stack and freerdp. What else can I check? And is there a way to specify explicitly which pulse device to use to make sure any kind of confusion is avoided?

Meanwhile I'll try to build freerdp with more codecs support to see if it helps anything in the ALSA mode (since I'm also hitting that creeping latency issue when used with that).

**Comment by shmerl** [comment]
OK, I managed to make it work in pulse mode for the microphone when using `/microphone:format:1` and it picked PCM. I also rebuilt all package with FAAC and FAAD2 support. How can I select those codecs? I tried using some other indexes for format, but that didn't work.

Like using `0xA106` for index for example (both in hexadecimal and in decimal notation). I.e. the number from:

```
[16:51:24:793] [162678:162744] [DEBUG][com.freerdp.channels.audin.client] - NumFormats 26
[16:51:24:793] [162678:162744] [DEBUG][com.freerdp.channels.audin.client] - WAVE_FORMAT_AAC_MS:	 wFormatTag: 0xA106 nChannels: 2
```

There is still the creeping latency issue even with format:1, it's just a bit better than situation with ALAW in Alsa mode. When running something like VLC for example, it starts lagging pretty hard eventually.
